### PR TITLE
Added keyboard "Signum 3.1" (by Troy Fletcher) and layout "layerlunacy" (by Klaus Weidinger).

### PR DESCRIPTION
**Pull request by dunkelziffer**
Worked on master branch. Is this a problem?

## Description

Added keyboard "Signum 3.1" (by Troy Fletcher).
Added layout "layerlunacy" (by Klaus Weidinger).
Provided a Python tool to generate the keymap.c from a more sane input file.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by dunkelziffer** [comment]
I implemented all requested changes from drashna and noroadsleft.
I contacted Troy Fletcher (the creator and vendor of the custom PCB) to specify his desired VENDOR_ID, etc., but he hasn't replied, yet. 
I put "Flet" and "Sig3" in leetspeek for VENDOR_ID and PRODUCT_ID and "troyfletcher" as Manufacturer.
As Troy's PCBs work with both Pro Micro and Teensy, I didn't specify a bootloader.

File: keyboards/signum31/config.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#define DEBOUNCING_DELAY 5
+
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* key combination for command */
+#define IS_COMMAND() ( \
```
Comment (drashna) [inline keyboards/signum31/config.h]: Could this be removed?

Comment (dunkelziffer) [reply keyboards/signum31/config.h]: Probably yes. I guess you mean till line 43?

I didn't write this code myself. Troy provided a .json file for "https://kbfirmware.com/".
I imported the file on that website, downloaded the .hex from there and then started editing it manually from that point onwards. All code that I didn't write myself might be autogenerated from that website, but even if it was provided indirectly from Troy via the .json file, it is at least 3 months old and might be deprecated by now.

**Comment by dunkelziffer** [comment]
You can merge either of the last two commits as you see fit.

**Comment by jceb** [comment]
>     * If there are going to be other revisions of this board, this keyboard should be named `keyboards/signum/31` or something similar.

I'll create a pull request for the Signum 3.0 in the near future. If nesting folder is a good idea let's do it now.

**Comment by noroadsleft** [comment]
> I'll create a pull request for the Signum 3.0 in the near future. If nesting folders is a good idea let's do it now.

Nesting folders is great, especially if you're dealing with revisions of the same keyboard, because it makes it simple to share code between revisions.

You could put code that applies to every revision in `keyboards/signum/`, and then `keyboards/signum/rev30/` and `keyboards/signum/rev31/` can have code specific to 3.0 and 3.1, respectively.

Some working examples of this:

- [Keeb.io Iris](/qmk/qmk_firmware/tree/master/keyboards/keebio/iris/)
- [KBDfans KBD75](/qmk/qmk_firmware/tree/master/keyboards/kbdfans/kbd75/)

**Comment by drashna** [comment]
@jceb Could you also specify the bootloader, as requested above?

**Comment by jceb** [comment]
Actually, the Signum 3.0 and 3.1 are two very different keyboards. I suggest that the following structure is created: `keyboards/signum3.0/{teensy,promicro}/` and `keyboards/signum3.1/{teensy,promicro}/`. This allows us to specify the bootloader for the two controllers and we have a clear separation between the different keyboards.

Is it okay to have a dot in the path name? Troy, the manufacturer of the keyboard, told me that he'd prefer that. If not the underscore would be a good replacement. Since Troy constantly develops new keyboards Signum 30 might come out some day and create a conflict.

**Comment by drashna** [comment]
@jackhumbert or @skullydazed is there any issues with using a dot/period in the keyboard name?

File: keyboards/signum31/keymaps/default/generate_km.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+from sortedcontainers import SortedDict
```
Comment (skullydazed) [inline keyboards/signum31/keymaps/default/generate_km.py]: This is an interesting script but I think it's out of scope for this PR. Could you pull this out for now and submit it as a separate PR? It should probably go into `/util`.

Comment (jceb) [reply keyboards/signum31/keymaps/default/generate_km.py]: The script is already part of qmk: `keyboards/signum/3_0/elitec/keymap/default/generate_km.py`. I've been working on it to reduce external dependencies and also to make it more generic. My feeling is that it needs lots more work to make it generic enough for other keyboards. If it's okay with you, I'll keep developing in the safety of the keymap folder before it gets exposed to the public. Maybe we can have an issue to collect ideas and round off the rough edges.

**Comment by drashna** [comment]
I think this PR is ready to go.  

There are a couple of commends that may need to be addressed first though.

File: keyboards/signum31/config.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xF1E7
+#define PRODUCT_ID      0x5163
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    troyfletcher
+#define PRODUCT         keyboard
```
Comment (noroadsleft) [inline keyboards/signum31/config.h]:
```suggestion
#define PRODUCT         Signum 3.1
```

File: keyboards/signum31/signum31.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT( \
+	K000, K001, K002, K003, K004, K005, K006, K007, K008, K009, \
```
Comment (noroadsleft) [inline keyboards/signum31/signum31.h]: Lines 6-11 should be edited to resemble the physical layout of the assembled keyboard, and the layer blocks...

https://github.com/qmk/qmk_firmware/blob/f4750687bda728ec486496f89c14ebf54d177c0e/keyboards/signum31/keymaps/default/keymap.c#L643-L649

... of the `keymap.c` files should be edited accordingly.

File: keyboards/signum31/config.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#define DIODE_DIRECTION COL2ROW
+
+/* number of backlight levels */
+
+#ifdef BACKLIGHT_PIN
+#define BACKLIGHT_LEVELS 0
+#endif
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCING_DELAY 5
```
Comment (noroadsleft) [inline keyboards/signum31/config.h]: This setting was renamed.

```suggestion
#define DEBOUNCE 5
```

**Comment by drashna** [comment]
Any updates on this?
